### ISSUE TITLE
The version of fsautocomplete.exe that ships on MELPA doesn't work on mono 4.2.1

### ISSUE THREAD

**Issue by nosami** [fsharp/emacs-fsharp-mode#25]
The version of fsautocomplete.exe that ships on MELPA doesn't work on mono 4.2.1... but fsautocomplete built from latest source is fine

```
Jasons-iMac:FsAutoComplete jason$ mono /Users/jason/.emacs.d/elpa/fsharp-mode-20150902.732/bin/fsautocomplete.exe
project "/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/FsAutoComplete.fsproj"
{"Kind":"error","Data":"Unknown tools version: '12.0' . Known versions: '2.0' '3.0' '3.5' '4.0'"}


Jasons-iMac:FsAutoComplete jason$ mono /Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/bin/Debug/fsautocomplete.exe
project "/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/FsAutoComplete.fsproj"
{"Kind":"project","Data":{"Project":"/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/FsAutoComplete.fsproj","Files":["/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/AssemblyInfo.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/Utils.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/Debug.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/Environment.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/BlockingQueueAgent.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/Options.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/Parser.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/TipFormatter.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/FileSystem.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/CompilerServiceInterface.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/CommandResponse.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/CommandInput.fs","/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/Program.fs"],"Output":"/Users/jason/src/fsharp/FsAutoComplete/FsAutoComplete/bin/Debug/fsautocomplete.exe","References":["../packages/FSharp.Compiler.Service/lib/net45/FSharp.Compiler.Service.dll","../packages/FSharp.Core/lib/net40/FSharp.Core.dll","../packages/NDesk.Options/lib/NDesk.Options.dll","../packages/Newtonsoft.Json/lib/net45/Newtonsoft.Json.dll","/Library/Frameworks/Mono.framework/Versions/4.2.1/lib/mono/4.5/System.Core.dll","/Library/Frameworks/Mono.framework/Versions/4.2.1/lib/mono/4.5/System.Numerics.dll","/Library/Frameworks/Mono.framework/Versions/4.2.1/lib/mono/4.5/System.Xml.dll","/Library/Frameworks/Mono.framework/Versions/4.2.1/lib/mono/4.5/System.dll","/Library/Frameworks/Mono.framework/Versions/4.2.1/lib/mono/4.5/mscorlib.dll"],"Framework":"v4.5"}}
```

**Comment by rneatherway** [comment]
Interesting. There aren't any changes to project cracking between the two versions. Could you use `monodis` to dump the assemblies they are referencing please? I'm not sure of the exact options right now.

**Comment by rneatherway** [comment]
```
~/.emacs.d/elpa/fsharp-mode-20150902.732 $ monodis --assemblyref bin/fsautocomplete.exe
AssemblyRef Table
1: Version=4.0.0.0
	Name=mscorlib
	Flags=0x00000000
	Public Key:
0x00000000: B7 7A 5C 56 19 34 E0 89
2: Version=1.4.0.1
	Name=FSharp.Compiler.Service
	Flags=0x00000000
	Zero sized public key
3: Version=4.3.0.0
	Name=FSharp.Core
	Flags=0x00000000
	Public Key:
0x00000000: B0 3F 5F 7F 11 D5 0A 3A
4: Version=4.0.0.0
	Name=System
	Flags=0x00000000
	Public Key:
0x00000000: B7 7A 5C 56 19 34 E0 89
5: Version=7.0.0.0
	Name=Newtonsoft.Json
	Flags=0x00000000
	Public Key:
0x00000000: 30 AD 4F E6 B2 A6 AE ED
6: Version=0.2.1.0
	Name=NDesk.Options
	Flags=0x00000000
	Zero sized public key
7: Version=2.0.0.0
	Name=mscorlib
	Flags=0x00000000
	Public Key:
0x00000000: B7 7A 5C 56 19 34 E0 89
```

**Comment by nosami** [comment]
Sure

Jasons-iMac:~ jason$ monodis --assemblyref ~/.emacs.d/elpa/fsharp-mode-20150902.732/bin/fsautocomplete.exe
AssemblyRef Table
1: Version=4.0.0.0
	Name=mscorlib
	Flags=0x00000000
	Public Key:
0x00000000: B7 7A 5C 56 19 34 E0 89
2: Version=1.4.0.1
	Name=FSharp.Compiler.Service
	Flags=0x00000000
	Zero sized public key
3: Version=4.3.0.0
	Name=FSharp.Core
	Flags=0x00000000
	Public Key:
0x00000000: B0 3F 5F 7F 11 D5 0A 3A
4: Version=4.0.0.0
	Name=System
	Flags=0x00000000
	Public Key:
0x00000000: B7 7A 5C 56 19 34 E0 89
5: Version=7.0.0.0
	Name=Newtonsoft.Json
	Flags=0x00000000
	Public Key:
0x00000000: 30 AD 4F E6 B2 A6 AE ED
6: Version=0.2.1.0
	Name=NDesk.Options
	Flags=0x00000000
	Zero sized public key
7: Version=2.0.0.0
	Name=mscorlib
	Flags=0x00000000
	Public Key:
0x00000000: B7 7A 5C 56 19 34 E0 89

Jasons-iMac:~ jason$ monodis --assemblyref ~/src/fsharp/FsAutoComplete/FsAutoComplete/bin/Release/fsautocomplete.exe
AssemblyRef Table
1: Version=4.0.0.0
	Name=mscorlib
	Flags=0x00000000
	Public Key:
0x00000000: B7 7A 5C 56 19 34 E0 89
2: Version=1.4.0.1
	Name=FSharp.Compiler.Service
	Flags=0x00000000
	Zero sized public key
3: Version=4.3.0.0
	Name=FSharp.Core
	Flags=0x00000000
	Public Key:
0x00000000: B0 3F 5F 7F 11 D5 0A 3A
4: Version=4.0.0.0
	Name=System
	Flags=0x00000000
	Public Key:
0x00000000: B7 7A 5C 56 19 34 E0 89
5: Version=7.0.0.0
	Name=Newtonsoft.Json
	Flags=0x00000000
	Public Key:
0x00000000: 30 AD 4F E6 B2 A6 AE ED
6: Version=0.2.1.0
	Name=NDesk.Options
	Flags=0x00000000
	Zero sized public key
7: Version=2.0.0.0
	Name=mscorlib
	Flags=0x00000000
	Public Key:
0x00000000: B7 7A 5C 56 19 34 E0 89

fwiw.... The MELPA version works fine with older versions of mono

**Comment by rneatherway** [comment]
Very strange... I wonder if that is because FCS is picking up the wrong version of the MSBuild assemblies. I'm not 100% sure how to check, although it should be possible using `mono --debug` or `mono --trace`. One thing to try is to check if you can build from the 0.23.1 tag and see if that works. I suggest that because there are no differences to do with MSBuild or project cracking between 0.23.1 and master, and I suspect that mono is doing something weird with assembly resolution.

**Comment by nosami** [comment]
Well, I just tried master version of fsautocomplete for now with mono 4.2.1 - but then there is a new issue

```
Full name: FsAutoComplete.Environment.tryFindFile" "Comment" ""))))
error in process filter: split-string: Wrong type argument: stringp, ((#s(hash-table size 65 test equal rehash-size 1.5 rehash-threshold 0.8 data ("Signature" "val getFile : (string -> VolatileFile option)" "Comment" ""))))
```

I do think that this is a mono issue though. I can get things working for now by changing the exec-path in emacs `  (setq exec-path (append '("/path/to/older/mono") exec-path))`

**Comment by rneatherway** [comment]
Right, the master version won't work with current fsharp-mode because of a backwards-incompatible change to the tooltip command. It's returned in a more semantic way now, which means the editor can format it itself in a suitable way. So I need to update the elisp correspondingly.

It would still be great to know if a compiled 0.23.1 works or not on your new mono if you have time.

**Comment by nosami** [comment]
Yes, you are correct. If I compile 0.23.1 using mono 4.2.1, everything works fine. Thanks!

**Comment by rneatherway** [comment]
Hmmpf! So mono has broken backwards compatibility! Could you zip up your compiled version and attach it here and I'll check it works for me (mono 4.0.2)? If it does then I can at least fix this problem by re-releasing fsharp-mode with fsautocomplete compiled with mono 4.2?

**Comment by nosami** [comment]
I couldn't attach it here, so I uploaded it to https://github.com/nosami/nosami.github.io/blob/master/fsautocomplete.zip?raw=true

It's compiled as Release

**Comment by rneatherway** [comment]
Thanks very much, that works. I'm a bit unhappy that mono is behaving this way, but if this easy change fixes it then I will go for this.

**Comment by nosami** [comment]
It's possibly a temporary problem, no idea - this is an alpha version of mono

**Comment by rneatherway** [comment]
@nosami I think you accidentally compiled the wrong tag. The version number seems to be 0.21.

**Comment by nosami** [comment]
I just had a look as I haven't touched that folder since the other day. It
was definitely 0.23.1

```
Jasons-iMac:FsAutoComplete jason$ git status
HEAD detached at 0.23.1
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)
```

**Comment by nosami** [comment]
I just compiled it again and pushed to the same location if you want to try.

**Comment by rneatherway** [comment]
Right, I will double-check, thanks!

**Comment by rneatherway** [comment]
OK, thanks. I had a stray fsautocomplete.zip in my Downloads folder and used the wrong one. That new version works fine too, so I've replaced the copy in MELPA with that. If you could upgrade in the next few days and confirm it works that would be great.

**Comment by nosami** [comment]
It works great. Thanks!

**Comment by TheDarkTrumpet** [comment]
I'm not sure if if this was broken again or not.  I am using the most recent version of mono 

TDTMBP:bin dthole$ mono --version
Mono JIT compiler version 4.3.0 (master/b7e2658 2015年 10月18日 日曜日 11時09分01秒 CDT)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	LLVM:          supported, not enabled.
	GC:            sgen
TDTMBP:bin dthole$ 


I tried the same solution here, although I couldn't find that branch.  I did download the zip file, and replaced my version, but it gave the same issue for me.

**Comment by rneatherway** [comment]
I think this is an issue in FCS, fixed recently: https://github.com/fsharp/FSharp.Compiler.Service/commit/3e2505ab235d9452d0031c4cb5c9050081f25072

I am updating FsAutoComplete to use that new version, once that is ready I will release a new version of fsharp-mode. I'll post here when that occurs (a couple of days probably).

**Comment by rneatherway** [comment]
OK @TheDarkTrumpet the new version of fsharp-mode is released, please let me know if that works for you.

**Comment by TheDarkTrumpet** [comment]
Just gave it a try, and it looks like it works now.  Thank you very much for the change.  I'm curious what changed, and eventually want to help on the project, so I'll dig in the diff this weekend.

**Comment by TheDarkTrumpet** [comment]
Oh, to clarify.  I tested through melpa specifically.  I doubt it matters much the source, but incase you cared, I figured I mention it.

**Comment by rneatherway** [comment]
Thanks a lot for testing! I'm pretty sure that the fix is down to that change to FSharp.Compiler.Service, which was required because Mono changed the xbuild toolset numbering for some reason.

**Comment by rneatherway** [comment]
Also @TheDarkTrumpet: if you have time your input on this project would be very welcome, I am happy to help people get started.

**Comment by nosami** [comment]
Working for me too. Thanks!

**Comment by drvink** [comment]
This seems to have broken again on mono 4.2.2. I tried building my own FsAutoComplete and that didn't fix it either.
